File: hw/otavalo/u1/hw09AntColonySimulator/AntColonySimulator/src/antcolonysimulator/model/Area.java
package antcolonysimulator.model;

/**
 *
 * @author Arelys Otavalo
 */

import java.util.Random; 

public class Area {
    private int width;
    private int height;
    private Cell[][] grid;
    private Colony[] colonies;
    private int colonyCount;
    private AntEater[] antEaters;
    private int eaterCount;
    private Random random = new Random(); 

    public Area(int width, int height) {
        this.width = width;
        this.height = height;
        this.grid = new Cell[height][width];
        this.colonies = new Colony[5];
        this.colonyCount = 0;
        this.antEaters = new AntEater[5];
        this.eaterCount = 0;
        setup();
    }

    private void setup() {
        for (int row = 0; row < height; row++)
            for (int col = 0; col < width; col++)
                grid[row][col] = new Cell(row, col, this);
    }

    public void addColony(Colony colony) {
        if (colonyCount < colonies.length)
            colonies[colonyCount++] = colony;
    }

    public void addAntEater(AntEater eater) {
        if (eaterCount < antEaters.length)
            antEaters[eaterCount++] = eater;
    }

    public Cell getCell(int row, int col) {
        if (row < 0 || row >= height || col < 0 || col >= width)
            return null;
        return grid[row][col];
    }
    
    private void trySpawnFood() {
        if (random.nextDouble() < 0.05) { 
            int row = random.nextInt(height);
            int col = random.nextInt(width);
            Cell cell = grid[row][col];
            
            if (cell.getFoodPile() == null) {
                int foodAmount = random.nextInt(41) + 10; 
                cell.setFoodPile(new FoodPile(cell, foodAmount));
                System.out.println(" New food (+" + foodAmount + ") appeared in (" + row + ", " + col + ")");
            }
        }
    }

    public int getTotalFoodInNests() {
        int totalFood = 0;
        for (int i = 0; i < colonyCount; i++) {
            if (colonies[i] != null) {
                totalFood += colonies[i].getNest().totalFood();
            }
        }
        return totalFood;
    }

    public int getTotalPheromoneLevel() {
        int totalPheromone = 0;
        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++) {
                PheromoneDrop p = grid[i][j].getPheromone();
                if (p != null) {
                    totalPheromone += p.getCurrentLevel();
                }
            }
        return totalPheromone;
    }

    public void runTick() {
        for (int i = 0; i < colonyCount; i++)
            if (colonies[i] != null) colonies[i].run();

        for (int i = 0; i < eaterCount; i++)
            if (antEaters[i] != null) antEaters[i].run();

        trySpawnFood();

        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++)
                grid[i][j].decreasePheromone();


        System.out.println("----- State of the area -----");
        System.out.println("Active colonies: " + colonyCount);
        System.out.println("Anteaters: " + eaterCount);
        System.out.println("Total ants: " + getTotalAnts());
        System.out.println("Food in nests: " + getTotalFoodInNests()); 
        System.out.println("Pheromone level: " + getTotalPheromoneLevel()); 
        System.out.println("----------------------------");
    }

    public int getTotalAnts() {
        int total = 0;
        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++)
                total += grid[i][j].getAntCount();
        return total;
    }
    
    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }
}
        

File: hw/medina/u1/hw09antSimulator/model/area.py
from model.cell import Cell

class Area:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.cells = [[Cell(x, y) for y in range(height)] for x in range(width)]
        self.colonies = []
        self.ant_eaters = []
    
    def get_cell(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.cells[x][y]
        return None
    
    def get_neighbor_cells(self, x, y):
        neighbors = []
        for dx, dy in [(0,1), (1,0), (0,-1), (-1,0), (1,1), (1,-1), (-1,1), (-1,-1)]:
            neighbor = self.get_cell(x + dx, y + dy)
            if neighbor:
                neighbors.append(neighbor)
        return neighbors
    
    def add_colony(self, position):
        from model.colony import Colony
        colony = Colony(position, self)
        self.colonies.append(colony)
        return colony
    
    def add_ant_eater(self, position):
        from model.ant_eater import AntEater
        ant_eater = AntEater(position, self)
        self.ant_eaters.append(ant_eater)
        self.get_cell(position[0], position[1]).add_ant_eater(ant_eater)
        return ant_eater
    
    def add_food_pile(self, x, y, amount):
        cell = self.get_cell(x, y)
        if cell:
            cell.add_food_pile(amount)
            print(f"Food pile added at ({x}, {y}) with {amount} food")
    
    def update(self):
        # Update all colonies
        for colony in self.colonies:
            colony.update()
        
        # Update all ant eaters
        for ant_eater in self.ant_eaters:
            ant_eater.act()
        
        # Update pheromones
        for x in range(self.width):
            for y in range(self.height):
                cell = self.cells[x][y]
                if cell.pheromone:
                    if not cell.pheromone.decrease_level():
                        cell.pheromone = None